nonprimary_foreignkey/fields.py

```diff
@@ -0,0 +1,111 @@
+import django
+from django.core import checks
+from django.utils.functional import cached_property
+
+if django.VERSION < (1, 8):
+    from django.db.models.loading import get_model
+else:
+    from django.apps import apps
+    get_model = apps.get_model
+
+
+UNDEFINED = object()
+
+
+class NonPrimaryForeignKey(object):
+    """
+    Descriptor class for handling non-primary foreign keys.
+
+    The implementation is loosely based on ``GenericForeignKey`` from the
+    ``django.contrib.contenttypes`` module.
+    """
+
+    def __init__(self, to_model, from_field, to_field):
+        self._to_model = to_model
+        self._from_field = from_field
+        self._to_field = to_field
+        self.editable = False
+
+    @cached_property
+    def to_model(self):
+        return get_model(self._to_model)
+
+    @cached_property
+    def to_field(self):
+        return self.to_model._meta.get_field(self._to_field)
+
+    @cached_property
+    def from_field(self):
+        return self.model._meta.get_field(self._from_field)
+
+    def contribute_to_class(self, cls, name):
+        self.name = name
+        self.model = cls
+        self.cache_name = "_%s_cache" % name
+        setattr(cls, name, self)
+
+    def __get__(self, instance, instance_type=None):
+        if instance is None:
+            return self
+
+        value = getattr(instance, self.from_field.attname, None)
+        if value is None:
+            return None
+        try:
+            cached = getattr(instance, self.cache_name)
+        except AttributeError:
+            cached = UNDEFINED
+        if cached is not UNDEFINED:
+            if getattr(cached, self.to_field.attname, None) == value:
```

Comment:
Nice!  I was wondering how you were going to handle this side of the cache invalidation problem.